### PR TITLE
fix torch multi-gpu training hanging and timeout problem 

### LINKED ISSUE #1454: multi-gpu training: training hangs after one sub-epoch
Multi-gpu training using 4 GPUs on one single node (`torch_distributed` set to `{"class": torch.nn.parallel.DistributedDataParallel, "options": {}})`: training hangs after one sub-epoch.

Training hanging issue, complete log file below:
```
Torch distributed initialized. Hostname gpu-12.apptek.local, pid 11642, rank 1 / size 4, local rank 1 / local size 4.
Torch distributed initialized. Hostname gpu-12.apptek.local, pid 11643, rank 2 / size 4, local rank 2 / local size 4.
RETURNN starting up, version 1.20231104.223301+git.b4dbaddd, date/time 2023-11-05-05-02-24 (UTC-0500), pid 11643, cwd /nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/returnn/training/ReturnnTrainingJob.SJfEN18jeHhc/work, Python /usr/bin/python3TURNN starting up, version 1.20231104.223301+git.b4dbaddd, date/time 2023-11-05-05-02-24 (UTC-0500), pid 11642, cwd /nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/returnn/training/ReturnnTrainingJob.SJfEN18jeHhc/work, Python /usr/bin/python3rch distributed initialized. Hostname gpu-12.apptek.local, pid 11641, rank 0 / size 4, local rank 0 / local size 4.
RETURNN starting up, version 1.20231104.223301+git.b4dbaddd, date/time 2023-11-05-05-02-24 (UTC-0500), pid 11641, cwd /nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/returnn/training/ReturnnTrainingJob.SJfEN18jeHhc/work, Python /usr/bin/python3rch distributed initialized. Hostname gpu-12.apptek.local, pid 11644, rank 3 / size 4, local rank 3 / local size 4.
RETURNN starting up, version 1.20231104.223301+git.b4dbaddd, date/time 2023-11-05-05-02-24 (UTC-0500), pid 11644, cwd /nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/returnn/training/ReturnnTrainingJob.SJfEN18jeHhc/work, Python /usr/bin/python3ExtModCompiler call: g++ -shared -O2 -std=c++11 -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -I /nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/frontend/_native -I /usr/include/python3.10 -fPIC -v -D_GLIBCXX_USE_CXX11_ABI=0 -g /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/_returnn_frontend_native.cc -o /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/_returnn_frontend_native.so
Reading sequence list for MetaDataset 'dev' from sub-dataset 'dev_alignments'
PyExtModCompiler call: g++ -shared -O2 -std=c++11 -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -I /nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/frontend/_native -I /usr/include/python3.10 -fPIC -v -D_GLIBCXX_USE_CXX11_ABI=0 -g /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/_returnn_frontend_native.cc -o /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/_returnn_frontend_native.so
Waiting for lock-file: /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/lock_file
Torch: Hostname gpu-12.apptek.local, pid 11643, using GPU 2.
PyTorch: 2.0.1+cu117 (e9ebda29d87ce0916ab08c06ab26fd3766a870e5) (<not-under-git> in /usr/local/lib/python3.10/dist-packages/torch)
Reading sequence list for MetaDataset 'dev' from sub-dataset 'dev_alignments'
[2023-11-05 10:02:36,720] INFO: Run time: 0:00:55 CPU: 0.40% RSS: 3.01GB VMS: 88.16GB
Train data:
  input: 0 x 1
  output: {'data': [1, 2], 'classes': [9001, 1]}
  MetaDataset, sequences: 1317274, frames: unknown
Dev data:
  MetaDataset, sequences: 6579, frames: unknown
PyExtModCompiler call: g++ -shared -O2 -std=c++11 -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -I /nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/frontend/_native -I /usr/include/python3.10 -fPIC -v -D_GLIBCXX_USE_CXX11_ABI=0 -g /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/_returnn_frontend_native.cc -o /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/_returnn_frontend_native.so
Using device: cuda ('gpu' in config)
Start running torch distributed training on local rank 3.
/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.lOKkFj0ksUUM/output/i6_models/i6_models/config.py:86: TypeHintWarning: Skipping type check against 'VGG4LayerActFrontendV1Config'; this looks like a string-form forward reference imported from another module
  typeguard.check_type(self.cfg, cfg_parameter.annotation)
Reading sequence list for MetaDataset 'train' from sub-dataset 'train_alignments'
net params #: 78693129
Train data:
  input: 0 x 1
  output: {'data': [1, 2], 'classes': [9001, 1]}
  MetaDataset, sequences: 1317274, frames: unknown
Dev data:
  MetaDataset, sequences: 6579, frames: unknown
Using device: cuda ('gpu' in config)
Start running torch distributed training on local rank 2.
/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.lOKkFj0ksUUM/output/i6_models/i6_models/config.py:86: TypeHintWarning: Skipping type check against 'VGG4LayerActFrontendV1Config'; this looks like a string-form forward reference imported from another module
  typeguard.check_type(self.cfg, cfg_parameter.annotation)
net params #: 78693129
Torch: Hostname gpu-12.apptek.local, pid 11642, using GPU 1.
PyTorch: 2.0.1+cu117 (e9ebda29d87ce0916ab08c06ab26fd3766a870e5) (<not-under-git> in /usr/local/lib/python3.10/dist-packages/torch)
Reading sequence list for MetaDataset 'dev' from sub-dataset 'dev_alignments'
PyExtModCompiler call: g++ -shared -O2 -std=c++11 -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -I /nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/frontend/_native -I /usr/include/python3.10 -fPIC -v -D_GLIBCXX_USE_CXX11_ABI=0 -g /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/_returnn_frontend_native.cc -o /tmp/3289467.1.gpu.q/pzheng/returnn_py_ext_mod_cache/_returnn_frontend_native/0f5f147c83/_returnn_frontend_native.so
[2023-11-05 10:02:41,740] INFO: Run time: 0:01:00 CPU: 0.20% RSS: 6.86GB VMS: 132.85GB
Reading sequence list for MetaDataset 'train' from sub-dataset 'train_alignments'
Train data:
  input: 0 x 1
  output: {'data': [1, 2], 'classes': [9001, 1]}
  MetaDataset, sequences: 1317274, frames: unknown
Dev data:
  MetaDataset, sequences: 6579, frames: unknown
Using device: cuda ('gpu' in config)
Start running torch distributed training on local rank 1.
/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.lOKkFj0ksUUM/output/i6_models/i6_models/config.py:86: TypeHintWarning: Skipping type check against 'VGG4LayerActFrontendV1Config'; this looks like a string-form forward reference imported from another module
  typeguard.check_type(self.cfg, cfg_parameter.annotation)
Torch: Hostname gpu-12.apptek.local, pid 11641, using GPU 0.
PyTorch: 2.0.1+cu117 (e9ebda29d87ce0916ab08c06ab26fd3766a870e5) (<not-under-git> in /usr/local/lib/python3.10/dist-packages/torch)
Reading sequence list for MetaDataset 'dev' from sub-dataset 'dev_alignments'
net params #: 78693129
Reading sequence list for MetaDataset 'train' from sub-dataset 'train_alignments'
[2023-11-05 10:02:46,756] INFO: Run time: 0:01:05 CPU: 0.40% RSS: 9.53GB VMS: 156.38GB
Train data:
  input: 0 x 1
  output: {'data': [1, 2], 'classes': [9001, 1]}
  MetaDataset, sequences: 1317274, frames: unknown
Dev data:
  MetaDataset, sequences: 6579, frames: unknown
Using device: cuda ('gpu' in config)
Start running torch distributed training on local rank 0.
/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.lOKkFj0ksUUM/output/i6_models/i6_models/config.py:86: TypeHintWarning: Skipping type check against 'VGG4LayerActFrontendV1Config'; this looks like a string-form forward reference imported from another module
  typeguard.check_type(self.cfg, cfg_parameter.annotation)
net params #: 78693129
[2023-11-05 10:02:51,777] INFO: Run time: 0:01:10 CPU: 0.20% RSS: 12.45GB VMS: 180.09GB
Optimizer: Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    lr: 1.0
    maximize: False
    weight_decay: 0
)
Starting training at epoch 1, global train step 0
Optimizer: Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    lr: 1.0
    maximize: False
    weight_decay: 0
)
Starting training at epoch 1, global train step 0
Optimizer: Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    lr: 1.0
    maximize: False
    weight_decay: 0
)
Starting training at epoch 1, global train step 0
Optimizer: Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    lr: 1.0
    maximize: False
    weight_decay: 0
)
Starting training at epoch 1, global train step 0
start epoch 1 global train step 0 with effective learning rate 1e-05 ...
start epoch 1 global train step 0 with effective learning rate 1e-05 ...
start epoch 1 global train step 0 with effective learning rate 1e-05 ...
start epoch 1 global train step 0 with effective learning rate 1e-05 ...
[2023-11-05 10:02:56,909] INFO: Run time: 0:01:15 CPU: 0.40% RSS: 17.02GB VMS: 193.15GB
[2023-11-05 10:04:22,223] INFO: Run time: 0:02:40 CPU: 0.20% RSS: 18.76GB VMS: 194.88GB
[2023-11-05 10:05:12,407] INFO: Run time: 0:03:30 CPU: 0.40% RSS: 22.50GB VMS: 203.38GB
[2023-11-05 10:05:27,460] INFO: Run time: 0:03:46 CPU: 0.20% RSS: 25.71GB VMS: 222.92GB
[2023-11-05 10:11:13,808] INFO: Run time: 0:09:32 CPU: 0.40% RSS: 28.28GB VMS: 230.39GB
[2023-11-05 10:33:48,563] INFO: Run time: 0:32:07 CPU: 0.40% RSS: 31.11GB VMS: 233.33GB
Trained 13301 steps, 0:47:07 elapsed (91.5% computing time)
Trained 13301 steps, 0:47:07 elapsed (91.0% computing time)
Trained 13301 steps, 0:47:07 elapsed (91.0% computing time)
Total train loss: ce 5.298 fer 0.735
Total train loss: ce 5.303 fer 0.736
Total train loss: ce 5.323 fer 0.738
start epoch 2 global train step 13301 with effective learning rate 3.111111111111111e-05 ...
start epoch 2 global train step 13301 with effective learning rate 3.111111111111111e-05 ...
start epoch 2 global train step 13301 with effective learning rate 3.111111111111111e-05 ...
Trained 13301 steps, 0:47:07 elapsed (90.9% computing time)
Total train loss: ce 5.313 fer 0.737
Evaluating dataset 'dev'
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/activation.py:1160: UserWarning: Converting mask without torch.bool dtype to bool; this will negatively affect performance. Prefer to use a boolean mask directly. (Triggered internally at ../aten/src/ATen/native/transformers/attention.cpp:150.)
  return torch._native_multi_head_attention(
dev: score 4.095 error 0.683
Only 1 epochs stored so far and keeping last 2 epochs and best 4 epochs, thus not cleaning up any epochs yet.
start epoch 2 global train step 13301 with effective learning rate 3.111111111111111e-05 ...
Unhandled exception <class 'RuntimeError'> in thread <_MainThread(MainThread, started 22977597855552)>, proc 11644.Unhandled exception <class 'RuntimeError'> in thread <_MainThread(MainThread, started 22911415002944)>, proc 11643.



Thread current, main, <_MainThread(MainThread, started 22911415002944)>:
(Excluded thread.)

That were all threads.Thread current, main, <_MainThread(MainThread, started 22977597855552)>:

(Excluded thread.)

That were all threads.
Unhandled exception <class 'RuntimeError'> in thread <_MainThread(MainThread, started 22558761436992)>, proc 11642.

Thread current, main, <_MainThread(MainThread, started 22558761436992)>:
(Excluded thread.)

That were all threads.
EXCEPTION
Traceback (most recent call last):
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/rnn.py", line 11, in <module>
    line: main()
    locals:
      main = <local> <function main at 0x14e5df55bf40>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/__main__.py", line 634, in main
    line: execute_main_task()
    locals:
      execute_main_task = <global> <function execute_main_task at 0x14e5df55be20>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/__main__.py", line 440, in execute_main_task
    line: engine.train()
    locals:
      engine = <global> <returnn.torch.engine.Engine object at 0x14e4f5ba9b70>
      engine.train = <global> <bound method Engine.train of <returnn.torch.engine.Engine object at 0x14e4f5ba9b70>>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/torch/engine.py", line 225, in Engine.train
    line: self.train_epoch()
    locals:
      self = <local> <returnn.torch.engine.Engine object at 0x14e4f5ba9b70>
      self.train_epoch = <local> <bound method Engine.train_epoch of <returnn.torch.engine.Engine object at 0x14e4f5ba9b70>>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/torch/engine.py", line 278, in Engine.train_epoch
    line: torch.distributed.all_reduce(_has_data, op=torch.distributed.ReduceOp.MIN)
    locals:
      torch = <global> <module 'torch' from '/usr/local/lib/python3.10/dist-packages/torch/__init__.py'>
      torch.distributed = <global> <module 'torch.distributed' from '/usr/local/lib/python3.10/dist-packages/torch/distributed/__init__.py'>
      torch.distributed.all_reduce = <global> <function all_reduce at 0x14e5bd5cc820>
      _has_data = <local> tensor([1], dtype=torch.int8)
      op = <not found>
      torch.distributed.ReduceOp = <global> <class 'torch.distributed.distributed_c10d.ReduceOp'>
      torch.distributed.ReduceOp.MIN = <global> <RedOpType.MIN: 3>
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/distributed_c10d.py", line 1451, in broadcast_multigpu
    line: return func(*args, **kwargs)
    locals:
      func = <local> <function all_reduce at 0x14e5bd5cc790>
      args = <local> (tensor([1], dtype=torch.int8),)
      kwargs = <local> {'op': <RedOpType.MIN: 3>}
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/distributed_c10d.py", line 1707, in all_reduce
    line: work.wait()
    locals:
      work = <local> <torch.distributed.distributed_c10d.Work object at 0x14e5defc5a70>
      work.wait = <local> <bound method PyCapsule.wait of <torch.distributed.distributed_c10d.Work object at 0x14e5defc5a70>>
RuntimeError: [../third_party/gloo/gloo/transport/tcp/unbound_buffer.cc:81] Timed out waiting 1800000ms for recv operation to complete
EXCEPTION
Traceback (most recent call last):
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/rnn.py", line 11, in <module>
    line: main()
    locals:
      main = <local> <function main at 0x14845a7040d0>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/__main__.py", line 634, in main
    line: execute_main_task()
    locals:
      execute_main_task = <global> <function execute_main_task at 0x14845ad5bf40>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/__main__.py", line 440, in execute_main_task
    line: engine.train()
    locals:
      engine = <global> <returnn.torch.engine.Engine object at 0x14843a112680>
      engine.train = <global> <bound method Engine.train of <returnn.torch.engine.Engine object at 0x14843a112680>>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/torch/engine.py", line 225, in Engine.train
    line: self.train_epoch()
    locals:
      self = <local> <returnn.torch.engine.Engine object at 0x14843a112680>
      self.train_epoch = <local> <bound method Engine.train_epoch of <returnn.torch.engine.Engine object at 0x14843a112680>>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/torch/engine.py", line 278, in Engine.train_epoch
    line: torch.distributed.all_reduce(_has_data, op=torch.distributed.ReduceOp.MIN)
    locals:
      torch = <global> <module 'torch' from '/usr/local/lib/python3.10/dist-packages/torch/__init__.py'>
      torch.distributed = <global> <module 'torch.distributed' from '/usr/local/lib/python3.10/dist-packages/torch/distributed/__init__.py'>
      torch.distributed.all_reduce = <global> <function all_reduce at 0x1484389c8940>
      _has_data = <local> tensor([1], dtype=torch.int8)
      op = <not found>
      torch.distributed.ReduceOp = <global> <class 'torch.distributed.distributed_c10d.ReduceOp'>
      torch.distributed.ReduceOp.MIN = <global> <RedOpType.MIN: 3>
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/distributed_c10d.py", line 1451, in broadcast_multigpu
    line: return func(*args, **kwargs)
    locals:
      func = <local> <function all_reduce at 0x1484389c88b0>
      args = <local> (tensor([1], dtype=torch.int8),)
      kwargs = <local> {'op': <RedOpType.MIN: 3>}
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/distributed_c10d.py", line 1707, in all_reduce
    line: work.wait()
    locals:
      work = <local> <torch.distributed.distributed_c10d.Work object at 0x14826f4276b0>
      work.wait = <local> <bound method PyCapsule.wait of <torch.distributed.distributed_c10d.Work object at 0x14826f4276b0>>
RuntimeError: [../third_party/gloo/gloo/transport/tcp/unbound_buffer.cc:133] Timed out waiting 1800000ms for send operation to complete
XCEPTION
Traceback (most recent call last):
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/rnn.py", line 11, in <module>
    line: main()
    locals:
      main = <local> <function main at 0x14d67694beb0>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/__main__.py", line 634, in main
    line: execute_main_task()
    locals:
      execute_main_task = <global> <function execute_main_task at 0x14d67694bd90>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/__main__.py", line 440, in execute_main_task
    line: engine.train()
    locals:
      engine = <global> <returnn.torch.engine.Engine object at 0x14d67911c700>
      engine.train = <global> <bound method Engine.train of <returnn.torch.engine.Engine object at 0x14d67911c700>>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/torch/engine.py", line 225, in Engine.train
    line: self.train_epoch()
    locals:
      self = <local> <returnn.torch.engine.Engine object at 0x14d67911c700>
      self.train_epoch = <local> <bound method Engine.train_epoch of <returnn.torch.engine.Engine object at 0x14d67911c700>>
  File "/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/returnn/torch/engine.py", line 278, in Engine.train_epoch
    line: torch.distributed.all_reduce(_has_data, op=torch.distributed.ReduceOp.MIN)
    locals:
      torch = <global> <module 'torch' from '/usr/local/lib/python3.10/dist-packages/torch/__init__.py'>
      torch.distributed = <global> <module 'torch.distributed' from '/usr/local/lib/python3.10/dist-packages/torch/distributed/__init__.py'>
      torch.distributed.all_reduce = <global> <function all_reduce at 0x14d6545bc790>
      _has_data = <local> tensor([1], dtype=torch.int8)
      op = <not found>
      torch.distributed.ReduceOp = <global> <class 'torch.distributed.distributed_c10d.ReduceOp'>
      torch.distributed.ReduceOp.MIN = <global> <RedOpType.MIN: 3>
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/distributed_c10d.py", line 1451, in broadcast_multigpu
    line: return func(*args, **kwargs)
    locals:
      func = <local> <function all_reduce at 0x14d6545bc700>
      args = <local> (tensor([1], dtype=torch.int8),)
      kwargs = <local> {'op': <RedOpType.MIN: 3>}
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/distributed_c10d.py", line 1707, in all_reduce
    line: work.wait()
    locals:
      work = <local> <torch.distributed.distributed_c10d.Work object at 0x14d490130870>
      work.wait = <local> <bound method PyCapsule.wait of <torch.distributed.distributed_c10d.Work object at 0x14d490130870>>
RuntimeError: [../third_party/gloo/gloo/transport/tcp/unbound_buffer.cc:81] Timed out waiting 1800000ms for recv operation to complete
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 11641 closing signal SIGTERM
ERROR:torch.distributed.elastic.multiprocessing.api:failed (exitcode: 1) local_rank: 1 (pid: 11642) of binary: /usr/bin/python3
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 346, in wrapper
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/run.py", line 794, in main
    run(args)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/run.py", line 785, in run
    elastic_launch(
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 250, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
/nas/models/asr/pzheng/returnn_pytorch_backend/work/i6_core/tools/git/CloneGitRepositoryJob.LClNkVyrEDKK/output/repository/rnn.py FAILED
------------------------------------------------------------
Failures:
[1]:
  time      : 2023-11-05_06:21:10
  host      : gpu-12.apptek.local
  rank      : 2 (local_rank: 2)
  exitcode  : 1 (pid: 11643)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
[2]:
  time      : 2023-11-05_06:21:10
  host      : gpu-12.apptek.local
  rank      : 3 (local_rank: 3)
  exitcode  : 1 (pid: 11644)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2023-11-05_06:21:10
  host      : gpu-12.apptek.local
  rank      : 1 (local_rank: 1)
  exitcode  : 1 (pid: 11642)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================

```

_Originally posted by @kuacakuaca in https://github.com/rwth-i6/returnn/issues/1452#issuecomment-1798098777_

### PR DESCRIPTION
**Pull request by kuacakuaca**
Fix #1454.

- get dev dataloader iterator for all training processes to allow synchronization
- remove unused imports

### CODE REVIEW

File: returnn/torch/engine.py [hunk 1]
```diff
@@ -3,15 +3,14 @@
 """
 
 from __future__ import annotations
-from typing import Optional, Any, Union, Callable, Dict
```
Comment (albertz) [inline returnn/torch/engine.py:6]: But `Callable` is used many times here?

File: returnn/torch/engine.py [hunk 2]
```diff
@@ -3,15 +3,14 @@
...
+from typing import Optional, Any, Union, Dict
 from contextlib import nullcontext
 
 import sys
 import gc
 import os
 import torch
 import time
-from torch.distributed import init_process_group
```
Comment (albertz) [inline returnn/torch/engine.py:14]: I really wonder, I think you are right, I also don't see any usage of this now, but why did PyCharm not complain about this?

File: returnn/torch/engine.py (hunk 1 above)
Comment (kuacakuaca) [reply returnn/torch/engine.py:6]: i only saw it in comments

Comment (albertz) [reply returnn/torch/engine.py:6]: See for example:
```
  returnn/torch/engine.py:74: WARNING PyUnresolvedReferencesInspection: Unresolved reference 'Callable'
  returnn/torch/engine.py:75: WARNING PyUnresolvedReferencesInspection: Unresolved reference 'Callable'
```

File: returnn/torch/engine.py (hunk 2 above)
Comment (albertz) [reply returnn/torch/engine.py:14]: Ah, but now, you see those errors:
```
  returnn/torch/engine.py:311: WARNING PyUnresolvedReferencesInspection: Cannot find reference 'distributed' in '__init__.py | __init__.py'
  returnn/torch/engine.py:311: WARNING PyUnresolvedReferencesInspection: Cannot find reference 'distributed' in '__init__.py | __init__.py'
  returnn/torch/engine.py:426: WARNING PyUnresolvedReferencesInspection: Cannot find reference 'distributed' in '__init__.py | __init__.py'
  returnn/torch/engine.py:491: WARNING PyUnresolvedReferencesInspection: Cannot find reference 'distributed' in '__init__.py | __init__.py'
```
So it means that `torch.distributed` is not available by default, you need to import it.

Comment (albertz) [reply returnn/torch/engine.py:14]: But just doing this should be enough:
```python
import torch.distributed
```

File: returnn/torch/engine.py (hunk 1 above)
Comment (albertz) [reply returnn/torch/engine.py:6]: > i only saw it in comments

Those are type information. It is used then. It must be imported.

**Review by albertz — APPROVED** [review]

**Pull request merged**